add model resources for CPMAnt (new)

pioliverse opened the pull request:
# What does this PR do?
<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->
Since the previous submission(#20711 ) had problems here and there, we have now resubmitted a new one.
## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts and @NielsRogge
- speech models: @sanchit-gandhi

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger and @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

pioliverse committed:
add models and tests

pioliverse commented:
> Thanks very much @pioliverse for iterating! I left a couple of comments, I think that some refactoring needs to be considered, after that we should be close to merge this! My main comments are:
> 
> * I think that you can wrap `CPMAntEmbedding` around a `nn.Embedding` layer even though scaling is needed. You can scale down after each call to the embedding module and make sure the input is scaled down before the `projection` call.
> * Make sure to inherit `CPMAntForCausalLM` from `CPMAntPreTrainedModel`, also make sure to follow the convention / good practices by checking what is done in OPT for instance: https://github.com/huggingface/transformers/blob/1543cee7c8c95ef47f832b1f37625ba2923c4994/src/transformers/models/opt/modeling_opt.py#L808
>    - this includes defining correctly a `lm_head` module, functions such as `get_input_embeddings`, `set_input_embeddings`, etc.
> * A lot of arguments from module's init seems to be unused, e.g. `init_std`. Try also to take the `config` object as a single argument from the init whenever possible (e.g. `CPMAntEncoder`)
> * Please make sure to follow the correct styling for docstrings (check my comments about that below)
> * If you have to initialize some weights with a specific distribution, try to initialize all the submodules weights inside `_init_weights` function from `CPMAntPreTrainedModel`
> * It's unclear to me why `forward` function is not defined in `CPMAntForCausalLM`
> * The code can be optimized here and there, I left some comments below on how you can achieve that
> * Please do not raise `RuntimeErrors` outside `if torch_is_available()`, otherwise `flax` & `tf` tests will fail
> Again thanks a lot for your efforts!

@younesbelkada Thanks for your patience in reviewing, I followed OPT convention and made the following changes:
> * `CPMAntEmbedding` and `CPMAntLinear` has been replaced by `nn.Embedding` and `nn.Linear` respectively.
> * `CPMAntForCausalLM` has been inherited from `CPMAntPreTrainedModel`, and `lm_head` and some functions have been added.
> * Useless initial arguments have been removed.
> * `forward` has been defined in `CPMAntForCausalLM`

pioliverse commented:
@younesbelkada Thanks again for your patience in reviewing.

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+        elif m_name == "kernel":
+            array = np.transpose(array)
+        try:
+            # assert (
+            #     pointer.shape == array.shape
+            # ), f"Pointer shape {pointer.shape} and array shape {array.shape} mismatched"
```
to clean up!

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+        Args:
+            x (`torch.Tensor` of shape `(batch, seq_len, dim_in)`)
+        """  # noqa: E501
+        # assert x.size(-1) == self.dim_norm
+        if x.size(-1) != self.dim_norm:
```
to clean up

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+    connection.
+
+        Args:
+            dim_model (int): Main dimension of modules in transformer blocks.
+            num_heads (int): Number of attention heads in the Transformer encoder.
+            dim_head (int): Dimension of attention heads for each attention layer in the Transformer encoder.
+            dtype (optional): Defaults to torch.float.
+            eps (float, optional): The epsilon used by the layer normalization layers.
+            dropout_p (float, optional): Defaults to 0.
```
Please respect the convention as above ;) 
```
dim_model (`int`):
    Main dimension of ...
...
```

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+    """The whole feed-forward block. A sequence of operation. Consists of layernorm, feed-forward and residual connection.
+
+    Args:
+        dim_model (int): Main dimension of modules in transformer blocks.
+        dim_ff (int): Dimension of the "intermediate" (i.e., feed-forward) layer in the Transformer encoder.
+        dtype (optional): Defaults to torch.float.
+        eps (float, optional): The epsilon used by the layer normalization layers.
+        dropout_p (float, optional): Defaults to 0.
```
Same comment as above!

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+    """Layers of encoder transformer blocks plus an final layernorm.
+
+    Args:
+        num_layers (int): Number of layers.
+        dim_model (int): Main dimension of modules in transformer blocks.
+        dim_ff (int): Dimension of the "intermediate" (i.e., feed-forward) layer in the Transformer encoder.
+        num_heads (int): Number of attention heads in the Transformer encoder.
+        dim_head (int): Dimension of attention heads for each attention layer in the Transformer encoder.
+        dtype (optional): Defaults to torch.float.
+        eps (float, optional): The epsilon used by the layer normalization layers.
+        dropout_p (float, optional): Defaults to 0.
```
same as above

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+            use_cache (`bool`): Whether use cache.
+            past_key_values (`Tuple(torch.FloatTensor)`, *optional*): Cached past key and value projection states.
+        """  # noqa: E501
+        if not use_cache:
+            for layer in self.layers:
```
can you upstream `use_cache` at the layer level? i.e. can you pass this argument in `CPMAntTransformerBlock` forward argument instead?

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+                hidden_states = layer(hidden_states, attention_mask, position_bias)
+            hidden_states = self.output_layernorm(hidden_states)
+            return hidden_states
+        with torch.no_grad():
```
why `torch.no_grad` here?

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+                    attention_mask,
+                    position_bias,
+                    past_key_value=past_key_values[i] if past_key_values else None,
+                    use_cache=use_cache,
```
I see that `use_cache` is also passed here, I am slightly confused why you don't pass it above too

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+            keylen = key_pos.size(1)
+            querylen = query_pos.size(1)
+
+            # assert key_pos.size(0) == query_pos.size(0)
+            # assert keylen == key_segment.size(1) and querylen == query_segment.size(1)
```
to clean up

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+        segment_states = self.segment_embedding(segment)
+        hidden_states = torch.cat([prompt_states, hidden_states], 1) + segment_states
+
+        with torch.no_grad():
+            device = input.device
+            directional_mask_2d = torch.arange(seqlen, device=device) <= torch.arange(seqlen, device=device).view(
+                -1, 1
+            )
+            attention_mask = context[:, None, :] | (
+                context[:, :, None].logical_not() & directional_mask_2d.view(1, seqlen, seqlen)
+            )
+            attention_mask = attention_mask & (span[:, None, :] == span[:, :, None])
+            mask_1d = torch.arange(seqlen, device=device)[None, :].repeat(batch, 1) < length[:, None]
+            attention_mask = mask_1d.view(batch, seqlen, 1) & mask_1d.view(batch, 1, seqlen) & attention_mask
```
nit: You can probably wrap that in a class method `_prepare_attention_mask`

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+
+        seqlen = past_length + input.size(1)
+
+        with torch.no_grad():
+            device = input.device
+            directional_mask_2d = torch.arange(seqlen, device=device) <= torch.arange(seqlen, device=device).view(
+                -1, 1
+            )
+            attention_mask = context[:, None, :] | (
+                context[:, :, None].logical_not() & directional_mask_2d.view(1, seqlen, seqlen)
+            )
+            attention_mask = attention_mask & (span[:, None, :] == span[:, :, None])
+            # mask for left paddding
+            mask_1d = (
+                torch.tensor(list(range(seqlen))[::-1], device=device)[None, :].repeat(batch, 1) < length[:, None]
+            )
+            attention_mask = mask_1d.view(batch, seqlen, 1) & mask_1d.view(batch, 1, seqlen) & attention_mask
```
same, you can wrap that in a method

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1475 @@
+            hidden_states, attention_mask, position_bias, True, past_key_values
+        )
+        logits = self.lm_head(hidden_states)
+        return logits, hidden_states, present_key_values
```
Can you return a dataclass object instead? For example: https://github.com/huggingface/transformers/blob/8fb4d0e4b46282d96386c229b9fb18bf7c80c25a/src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py#L697 or you can also define your own class

younesbelkada reviewed (commented):
Thanks so much for your patience ! Looks pretty clean thank you! We should be close merging this once most of the comments are addressed. My comments being:
### Docstring and comments:
- please harmonize the function docstrings to match the convention `transformers` model follow
- please make sure to clean up some comments
- Also would be nice to add a small explanation on the code on why `generate` needs to be overriden

### `dtype`:
- I don't think the argument `dtype` is needed. The dtype of the whole model is managed by the kwarg `torch_dtype` so you can load your model using `model = xxxForCausalLM.from_pretrained(xxx, torch_dtype=torch.float16)` or `torch_dtype="auto"`  (if the weights are pushed in fp16) and the model will be loaded in the desired precision.

### tests
- I think that a test is failing, please double check that

### general comments
- For classes that are public (i.e. that are ported in `__init__.py`, basically`CPMAntModel` &. `CPMAntForCausalLM` it is preferable to adopt this logic: https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py#L693-L703 --> outputa tuple if not `return_dict` otherwise return a dataclass. Please check other modeling files as reference 
- you can wrap `attention_mask` creation process inside class methods, e.g. `_prepare_attention_mask(`

Thanks!

pioliverse committed:
solve conflicts

pioliverse committed:
test

pioliverse committed:
reformat

pioliverse committed:
resolve conflicts

pioliverse committed:
test

pioliverse committed:
resolve conflicts

pioliverse committed:
test

pioliverse committed:
resolve conflicts

pioliverse committed:
resolve conflicts

pioliverse committed:
resolve conflicts

pioliverse committed:
resolve conflicts

pioliverse committed:
resolve conflicts

pioliverse committed:
resolve conflicts

pioliverse committed:
rebase and make style

pioliverse committed:
rebase and make style

pioliverse committed:
resolve conflicts

pioliverse committed:
resolve conflicts

pioliverse committed:
tests

pioliverse committed:
resolve conflicts

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20906). All of your documentation changes will be reflected on that endpoint.

pioliverse committed:
tests

pioliverse committed:
tests

pioliverse committed:
tests

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20906). All of your documentation changes will be reflected on that endpoint.

pioliverse committed:
resolve conflicts

pioliverse committed:
rewrite some functions

pioliverse committed:
resolve conflicts

pioliverse committed:
resolve conflicts

pioliverse commented:
> Thanks so much for your patience ! Looks pretty clean thank you! We should be close merging this once most of the comments are addressed. My comments being:
> 
> ### Docstring and comments:
> * please harmonize the function docstrings to match the convention `transformers` model follow
> * please make sure to clean up some comments
> * Also would be nice to add a small explanation on the code on why `generate` needs to be overriden
> 
> ### `dtype`:
> * I don't think the argument `dtype` is needed. The dtype of the whole model is managed by the kwarg `torch_dtype` so you can load your model using `model = xxxForCausalLM.from_pretrained(xxx, torch_dtype=torch.float16)` or `torch_dtype="auto"`  (if the weights are pushed in fp16) and the model will be loaded in the desired precision.
> 
> ### tests
> * I think that a test is failing, please double check that
> 
> ### general comments
> * For classes that are public (i.e. that are ported in `__init__.py`, basically`CPMAntModel` &. `CPMAntForCausalLM` it is preferable to adopt this logic: https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt_neox_japanese/modeling_gpt_neox_japanese.py#L693-L703 --> outputa tuple if not `return_dict` otherwise return a dataclass. Please check other modeling files as reference
> * you can wrap `attention_mask` creation process inside class methods, e.g. `_prepare_attention_mask(`
> 
> Thanks!



Hi @younesbelkada, we have made some changes as follows:
1. add some docstrings.
2. modified `forward` following the style of transformers.
3. rewrote some functions to adapt the `generate` function
>* in `modeling_cpmant.py`, we rewrote some functions like `prepare_inputs_for_generation`, `_expand_inputs_for_generation`
>* in `tokenization_cpmant.py`, rewrote some functions like `prepare_for_model`, `_pad`, `_encode_plus`, `_batch_encode_plus`
4. cleaned some comments.

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        if self.dropout is not None:
+            score = self.dropout(score)
+
+        # (b, n_h, len_q, len_k) @ (b, n_h, len_k, d_h) -> (b, n_h, len_q, d_h)
```
same comment as above

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+    connection.
+
+        Args:
+            dim_model (int): Main dimension of modules in transformer blocks.
+            num_heads (int): Number of attention heads in the Transformer encoder.
+            dim_head (int): Dimension of attention heads for each attention layer in the Transformer encoder.
+            eps (float, optional): The epsilon used by the layer normalization layers.
+            dropout_p (float, optional): Defaults to 0.
+            use_cache (bool, optional): Whether to use cache.
```
```suggestion
            dim_model (int): 
                Main dimension of modules in transformer blocks.
            num_heads (int): 
                Number of attention heads in the Transformer encoder.
            dim_head (int):
                Dimension of attention heads for each attention layer in the Transformer encoder.
            eps (float, optional): 
                The epsilon used by the layer normalization layers.
            dropout_p (float, optional): 
                Defaults to 0.
            use_cache (bool, optional): 
                Whether to use cache.
```

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        use_cache: Optional[bool] = False,
+    ):
+        super().__init__()
+        self.use_cache = use_cache
+        self.layernorm_before_attention = CPMAntLayerNorm(
+            dim_norm=dim_model,
+            eps=eps,
+        )
+        self.self_attention = CPMAntAttention(
+            dim_model=dim_model,
+            num_heads=num_heads,
+            dim_head=dim_head,
+            dropout_p=dropout_p,
+            use_cache=self.use_cache,
+        )
```
you can pass `use_cache` directly to the `forward` methods, that is usually how we do it in `transformers`

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        """  # noqa: E501
+        x = self.layernorm_before_attention(hidden_states)
+        x = self.self_attention(x, x, attention_mask, position_bias, past_key_value)
+        if self.use_cache:
+            x, current_key_value = x
```
here pass `use_cache` directly on the `forward` pass

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        self.w_1 = nn.Linear(dim_in, dim_ff, bias=False)
+        self.act = torch.nn.GELU()
+
+    def forward(self, x: torch.Tensor):
+        """Transform an input tensor from one feature space to another via a nonlinear operation
+
+        Args:
+            x (`torch.Tensor` of shape `(batch, seq_len, dim_in)`)
+        """  # noqa: E501
+        gate_score = self.act(self.w_0(x))
+        x = self.w_1(x)
+
+        x = gate_score * x
+        return x
```
```suggestion
    def forward(self, hidden_states: torch.Tensor):
        """Transform an input tensor from one feature space to another via a nonlinear operation

        Args:
            hidden_states (`torch.Tensor` of shape `(batch, seq_len, dim_in)`)
        """  # noqa: E501
        gate_score = self.act(self.w_0(hidden_states))
        hidden_states = self.w_1(hidden_states)

        hidden_states = gate_score * hidden_states
        return hidden_states
```

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+class CPMAntFFNBlock(nn.Module):
+    """The whole feed-forward block. A sequence of operation. Consists of layernorm, feed-forward and residual connection.
+
+    Args:
+        dim_model (int): Main dimension of modules in transformer blocks.
+        dim_ff (int): Dimension of the "intermediate" (i.e., feed-forward) layer in the Transformer encoder.
+        eps (float, optional): The epsilon used by the layer normalization layers.
+        dropout_p (float, optional): Defaults to 0.
```
Please try to use `transformers` convention for docstrings

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+            hidden_states (`torch.Tensor` of shape `(batch, len_seq, dim_model)`):
+                Hidden states before feed forward layer.
+        """  # noqa: E501
+        x = self.layernorm_before_ffn(hidden_states)
+        x = self.ffn(x)
+        if self.dropout is not None:
+            x = self.dropout(x)
+        hidden_states = hidden_states + x
```
please avoid using single variable names here, consider using another name, e.g. `layer_output`

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+    feed-forward block.
+
+        Args:
+            dim_model (int): Main dimension of modules in transformer blocks.
+            dim_ff (int): Dimension of the "intermediate" (i.e., feed-forward) layer in the Transformer encoder.
+            num_heads (int): Number of attention heads in the Transformer encoder.
+            dim_head (int): Dimension of attention heads for each attention layer in the Transformer encoder.
+            eps (float, optional): The epsilon used by the layer normalization layers.
+            dropout_p (float, optional): Defaults to 0.
+            use_cache (bool, optional): Whether to use cache.
```
same comment as above, if you want to comment this class, consider using the `transformers` convention

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        super().__init__()
+        self.mask_att = mask_att
+        self.mask_ffn = mask_ffn
+        self.use_cache = use_cache
```
Why you need to store it here as a class attribute?

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+
+    def __init__(
+        self,
+        num_layers: int = 48,
+        dim_model: int = 4096,
+        dim_ff: int = 10240,
+        num_heads: int = 32,
+        dim_head: int = 128,
+        eps: float = 1e-6,
```
these values can be retrieved directly from the config

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        return hidden_states, current_key_values
+
+
+class CPMAntIntermediate(nn.Module):
```
This seems to be copied from `BertIntermediate`: https://github.com/huggingface/transformers/blob/6c62cfb2eff095c181481d8ae86c7f836b65d2d7/src/transformers/models/bert/modeling_bert.py#L439 can you add a `# Copied from` statement?

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        return relative_buckets
+
+
+class CPMAntOutput(nn.Module):
```
This seems to be copied from `BertOutput`: https://github.com/huggingface/transformers/blob/6c62cfb2eff095c181481d8ae86c7f836b65d2d7/src/transformers/models/bert/modeling_bert.py#L454

pioliverse replied:
Yes, we added a statement.

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        position_bias = self.position_bias(position, position, segment, segment)
+
+        hidden_states = self.encoder(hidden_states, attention_mask, position_bias)
+        logits = F.linear(hidden_states, self.input_embedding.weight)
```
can't you call `self.input_embedding` directly?

pioliverse replied:
Maybe not, `self.input_embedding` works for `input_ids`, but not for `hidden_states`.

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        if not return_dict:
+            return tuple(v for v in [logits, hidden_states] if v is not None)
+
+        return BaseModelOutput(hidden_states=hidden_states)
```
some attributes are missing such as the attention outputs etc.

pioliverse replied:
We added some attributes.

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        context: Optional[torch.Tensor] = None,
+        position: Optional[torch.Tensor] = None,
+        segment: Optional[torch.Tensor] = None,
+        span: Optional[torch.Tensor] = None,
+        return_dict: Optional[bool] = False,
+        **kwargs,
```
keyword arguments that are needed for public models such as `output_attentions` seem to be missing here

pioliverse replied:
We added the `output_attentions `.

younesbelkada reviewed (commented):
Thanks a lot for addressing most of the comments of the previous review! And thank you for your huge work on refactoring the modeling script
I left some comments, mostly nits that can be solved easily. Note that for arguments such as `use_cache` etc, we prefer to pass them through the forward pass rather than setting them as a class attribute. 
Also, please consider passing a `CPMAntConfig` for the classes that have several attributes such as `CPMAntEncoder` 
Make sure also to correctly pass the required keyword arguments such as `past_key_values`, `output_attentions` etc, that are crucial for caching mechanism. You can check how this is done in OPT for example
Finally, the naming convention in `transformers` has changed a bit, we prefer to name models with a single capital letter (i.e. here `CPMAnt -> Cpmant`)
Again thanks for your efforts on this! Once the comments being solved, we should be very close merging this!

pioliverse committed:
rebase and make style

pioliverse committed:
resolve conflicts

pioliverse committed:
resolve conflicts

pioliverse committed:
resolve conflicts

pioliverse commented:
We found `load_tf_weights_in_bert ` is unused, so we have deleted.

> Thanks a lot for addressing most of the comments of the previous review! And thank you for your huge work on refactoring the modeling script I left some comments, mostly nits that can be solved easily. Note that for arguments such as `use_cache` etc, we prefer to pass them through the forward pass rather than setting them as a class attribute. Also, please consider passing a `CPMAntConfig` for the classes that have several attributes such as `CPMAntEncoder` Make sure also to correctly pass the required keyword arguments such as `past_key_values`, `output_attentions` etc, that are crucial for caching mechanism. You can check how this is done in OPT for example Finally, the naming convention in `transformers` has changed a bit, we prefer to name models with a single capital letter (i.e. here `CPMAnt -> Cpmant`) Again thanks for your efforts on this! Once the comments being solved, we should be very close merging this!



Thanks for your review, we have modified some code.
>*We pass the `use_cache` in `forward` function from a class attribute.
>*We simplify the code for the class attribute assignment and replace it with `CPMAntconfig`.
>*We added `past_key_values` and `output_attentions` in `forward` of CPMAntModel.
>*I kind of wonder if all files that contain the name `CPMAnt` should be changed to `Cpmant`?

pioliverse commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+        Args:
+            hidden_q (`torch.Tensor`):
+                Indices of input sequence tokens of shape `(batch, len_q, dim_model)`. It will be embedded by model's
+                internal embedding lookup matrix.
```
Revised.

pioliverse commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,1090 @@
+
+class CPMAntLayerNorm(nn.Module):
+    """RMS LayerNorm"""
+
```
We added some comments.

pioliverse committed:
fix load_tf_weights_in_cpmant

pioliverse committed:
fix load_tf_weights_in_cpmant

pioliverse committed:
fix load_tf_weights_in_cpmant

pioliverse committed:
fix load_tf_weights_in_cpmant

pioliverse committed:
reformat some unrelated files

pioliverse committed:
reformat some unrelated files

pioliverse committed:
reformat some unrelated files

pioliverse committed:
reformat some unrelated files

pioliverse commented:
I was a bit surprised that when I used `make style` command, some other files were reformatted as well.

pioliverse commented:
I am a bit surprised that when I use `make style`, some other files are also reformatted, which causes `check_code_quality` to fail.

younesbelkada commented:
Hi @pioliverse 
You need to rebase with `main` branch as the styling has been updated for most of the files in `transformers` , and update your black version as follows:
```
pip install --upgrade -e .["quality"]
```
Then `make style` or `make fixup`

pioliverse committed:
upgrade quality

pioliverse committed:
upgrade quality

pioliverse committed:
upgrade quality

pioliverse committed:
upgrade quality

pioliverse commented:
> Hi @pioliverse You need to rebase with `main` branch as the styling has been updated for most of the files in `transformers` , and update your black version as follows:
> 
> ```
> pip install --upgrade -e .["quality"]
> ```
> 
> Then `make style` or `make fixup`

Thanks @younesbelkada , this has been solved.

pioliverse commented:
> Thanks a lot for addressing most of the comments of the previous review! And thank you for your huge work on refactoring the modeling script I left some comments, mostly nits that can be solved easily. Note that for arguments such as `use_cache` etc, we prefer to pass them through the forward pass rather than setting them as a class attribute. Also, please consider passing a `CPMAntConfig` for the classes that have several attributes such as `CPMAntEncoder` Make sure also to correctly pass the required keyword arguments such as `past_key_values`, `output_attentions` etc, that are crucial for caching mechanism. You can check how this is done in OPT for example Finally, the naming convention in `transformers` has changed a bit, we prefer to name models with a single capital letter (i.e. here `CPMAnt -> Cpmant`) Again thanks for your efforts on this! Once the comments being solved, we should be very close merging this!



Thanks for your review @younesbelkada , we have modified some code.
>* We pass the `use_cache` in `forward` function from a class attribute.
>* We simplify the code for the class attribute assignment and replace it with `CPMAntconfig`.
>* We added `past_key_values` and `output_attentions` in `forward` of CPMAntModel.
>* I kind of wonder if all files that contain the name `CPMAnt` should be changed to `Cpmant`?

gongbaitao committed:
fix some bugs & docstring

gongbaitao committed:
resolve conflicts

gongbaitao committed:
resolve conflicts

gongbaitao committed:
fix some bugs & docstring

gongbaitao committed:
resolve conflicts

gongbaitao committed:
rebase and make style

gongbaitao committed:
save resolution

gongbaitao committed:
make style

gongbaitao committed:
make style

gongbaitao committed:
delete redefinition code

gongbaitao committed:
reformat function

gongbaitao commented:
Hi @younesbelkada , I rebase `pioliverse:cpmantmodel` with `huggingface:main`, but cannot pass the test. It seems the model `altclip` has some problems.
in tests_onnx I met the error:
<img width="786" alt="截屏2023-02-21 23 23 06" src="https://user-images.githubusercontent.com/45178523/220386456-4b0c04ef-2f0d-4a46-acca-ab730015d95a.png">
in tests_hub,
<img width="792" alt="截屏2023-02-21 23 25 11" src="https://user-images.githubusercontent.com/45178523/220386698-b30f5530-d350-4f72-819c-b58e5f611541.png">
How can I avoid such error?

gongbaitao committed:
fix bugs and refactor

gongbaitao committed:
modify docstrings and make style

younesbelkada commented:
Hi @gongbaitao 
Thanks for jumping in! And sorry for the delay
Rebasing with `main` should be probably solve this issue, will look into the PR asap, let me know once you think this is ready for review!

gongbaitao commented:
> Hi @gongbaitao Thanks for jumping in! And sorry for the delay Rebasing with `main` should be probably solve this issue, will look into the PR asap, let me know once you think this is ready for review!

Hi @younesbelkada , thanks for your reply and advice! 
In the latest commit, we refactor the code and make it as simple as possible. And I also rebase with `main`, but the error still exists. 
`ERROR tests/models/altclip/test_modeling_altclip.py
============ 72 passed, 551 skipped, 42 warnings, 1 error in 29.03s ============

Exited with code exit status 1
CircleCI received exit code 1`

gongbaitao committed:
unify import format in __init__.py

gongbaitao committed:
fix import-altclp bug

gongbaitao committed:
fix copies to update index.md

gongbaitao committed:
fix unused config parameters

gongbaitao committed:
fix unused config parameters

gongbaitao committed:
fix unused config parameters

gongbaitao commented:
Hi @younesbelkada, thanks for your reply and advice!
All the problems in unit test have beed solved. In the latest commit, we largely refactor the code to make it clear and simple. Hope you can give a review soon!

younesbelkada commented:
Thanks a lot for the heads-up @pioliverse @gongbaitao !
Quickly looking at the README_ja file it seems that a some unnecessarly changes were made, I suggest you merge this branch with the upstream `transformers` `main` branch:
```
git remote add upstream https://github.com/huggingface/transformers.git
git fetch upstream
git merge upstream/main
git push
```

I'll have a closer look on the other files asap!

gongbaitao committed:
Merge remote-tracking branch 'upstream/main' into cpmantmodel

gongbaitao committed:
update README_ja.md

gongbaitao committed:
dummy commit for unit test

gongbaitao commented:
> Thanks a lot for the heads-up @pioliverse @gongbaitao ! Quickly looking at the README_ja file it seems that a some unnecessarly changes were made, I suggest you merge this branch with the upstream `transformers` `main` branch:
> 
> ```
> git remote add upstream https://github.com/huggingface/transformers.git
> git fetch upstream
> git merge upstream/main
> git push
> ```
> 
> I'll have a closer look on the other files asap!

Thanks for the tips @younesbelkada!
 I have merged with `huggingface main` branch and update the README_ja, and it's up-to-date with the `main` now. Please kindly have a look.

gongbaitao committed:
fix attention mask

younesbelkada commented on tests/models/cpmant/test_modeling_cpmant.py:
```diff
@@ -0,0 +1,234 @@
+
+
+@require_torch
+class CPMAntModelTest(unittest.TestCase):
```
You need this class to be inherited from `ModelTesterMixin`, otherwise the common tests won't be run
You need to import it from: `from ...test_modeling_common import ModelTesterMixin`
Also, you need a separate tester for `xxxForCausalLM` class as the tester needs to be inherited from `GenerationTesterMixin`
Check [here](https://github.com/huggingface/transformers/blob/b29e2dcaff114762e65eaea739ba1076fc5d1c84/tests/models/gptsan_japanese/test_modeling_gptsan_japanese.py#L162) for more examples.

gongbaitao replied:
solved

younesbelkada reviewed (commented):
Thanks a lot for working on this and your hard work 💪 ! 
I left a couple of comments

1- My first question being about the availability of the public checkpoint on the Hub, I can't see `openbmb/cpm-ant-10b` on the Hub, the only model I see is: https://huggingface.co/oliverpi/old-cpm-ant-10b - you need to make sure the relevant weights are opensource to merge this PR! 

2- Make sure to add the `CPMAntTokenizer` & `CPMAntTokenizerFast` on the tokenizer auto-mapping

3- Make sure to change the logic behind `use_cache` on `CPMAntLayer` as I believe there is a much simpler solution!

4- Regarding slow integration tests, we might need to totally skip them as the weights are 40GB in fp32 (and 20GB in fp16), I think our daily CI runners have 16GB GPU VRAM, so friendly pinging here @ydshieh to see what could be the alternative. I believe this is also the reason why it is hard to have doctests here, but you might need to add doctests on the public classes (check an example [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gptsan_japanese/modeling_gptsan_japanese.py#L1151-L1200) )

5- You need to make sure the testing classes inherit from the relevant super-classes (i.e. `ModelTesterMixin`, `GenerationTesterMixin`) otherwise the common tests won't be run

Apart from that the code looks much more cleaner, the other comments are mostly nits! Again great work on the PR and looking forward to merging this!

gongbaitao committed:
add CPMAntTokenizer&-Fast to auto-mapping

gongbaitao committed:
drop redundant changes in README_ko

gongbaitao committed:
fix  defaults in docstring

gongbaitao committed:
fix use_cache and some docstring

gongbaitao committed:
add missing args in tokenizer

gongbaitao committed:
modify tester inheritance

gongbaitao committed:
add is_jieba_available

ydshieh commented:
> Regarding slow integration tests, we might need to totally skip them as the weights are 40GB in fp32 (and 20GB in fp16), I think our daily CI runners have 16GB GPU VRAM, so friendly pinging here @ydshieh to see what could be the alternative.

Yeah, just skip it/them.

gongbaitao committed:
fix some bugs

gongbaitao committed:
make style and fix-copies

gongbaitao committed:
add doctests

gongbaitao committed:
Merge remote-tracking branch 'upstream/main' into cpmantmodel

gongbaitao commented on README_ko.md:
```diff
@@ -262,8 +263,8 @@ Flax, PyTorch, TensorFlow 설치 페이지에서 이들을 conda로 설치하는
-1. **[EfficientFormer](https://huggingface.co/docs/transformers/model_doc/efficientformer)** (from Snap Research) released with the paper [EfficientFormer: Vision Transformers at MobileNetSpeed](https://arxiv.org/abs/2206.01191) by Yanyu Li, Geng Yuan, Yang Wen, Ju Hu, Georgios Evangelidis, Sergey Tulyakov, Yanzhi Wang, Jian Ren.
-1. **[EfficientNet](https://huggingface.co/docs/transformers/main/model_doc/efficientnet)** (from Google Brain) released with the paper [EfficientNet: Rethinking Model Scaling for Convolutional Neural Networks](https://arxiv.org/abs/1905.11946) by Mingxing Tan, Quoc V. Le.
+1. **[EfficientFormer](https://huggingface.co/docs/transformers/model_doc/efficientformer)** (Snap Research 에서 제공)은 Yanyu Li, Geng Yuan, Yang Wen, Ju Hu, Georgios Evangelidis, Sergey Tulyakov, Yanzhi Wang, Jian Ren.의 [EfficientFormer: Vision Transformers at MobileNetSpeed](https://arxiv.org/abs/2206.01191)논문과 함께 발표했습니다.
+1. **[EfficientNet](https://huggingface.co/docs/transformers/main/model_doc/efficientnet)** (Google Brain 에서 제공)은 Mingxing Tan, Quoc V. Le.의 [EfficientNet: Rethinking Model Scaling for Convolutional Neural Networks](https://arxiv.org/abs/1905.11946)논문과 함께 발표했습니다.
```
solved in be059963cacc06371fc20737bdcc9fa94987f104

gongbaitao commented on src/transformers/models/cpmant/configuration_cpmant.py:
```diff
@@ -0,0 +1,121 @@
+        segment_types (`int`, *optional*, defaults to 32):
+            The type of segment.
+        use_cache (`bool`, *optional*, defaults to 32):
+            Whether to use cache.
```
solved

gongbaitao committed:
skip integration tests

gongbaitao committed:
add is_jieba_available

gongbaitao committed:
fix bugs in common tests

gongbaitao committed:
Merge remote-tracking branch 'upstream/main' into cpmantmodel

gongbaitao committed:
adjust docstrings and make style

gongbaitao committed:
add argument docstring

gongbaitao commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,942 @@
+]
+
+
+# Adapted from Bert
```
solved

gongbaitao commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,942 @@
+    old_dtype = hidden.dtype
+    variance = hidden.to(torch.float32).pow(2).mean(dim=-1, keepdim=True)
+    hidden = (hidden * torch.rsqrt(variance + eps)).to(old_dtype)
+    return hidden * weight
```
solved

gongbaitao commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,942 @@
+    ):
+        """
+        Args:
+            hidden_q (`torch.Tensor`):
+                Input of transformer block(self-attention block). It can be the raw embedding of a batch of sequences.
+            hidden_kv (`torch.Tensor` of shape `(batch, len_k, dim_model)`)):
+                Tensor *key_value* and *query* of shape `(batch, len_k, dim_model)`
+            attention_mask (`torch.Tensor` of shape `(batch, len_seq, len_seq)`):
+                Avoid invalid areas to participate in the calculation of self-attention.
+            position_bias (`torch.Tensor` of shape `(batch, len_seq, len_seq)`):
+                Provide positional information to self-attention block.
+            output_attentions (`bool`, *optional*):
+                Whether or not to return the attentions tensors of all attention layers.
+            past_key_values (`Tuple[torch.Tensor, torch.Tensor]`, *optional*):
+                Cached past key and value projection states.
+            use_cache (`bool`, *optional*):
+                If set to `True`, `past_key_values` key value states are returned and can be used to speed up decoding
+                (see `past_key_values`).
+        """
```
solved

gongbaitao commented on src/transformers/models/cpmant/tokenization_cpmant.py:
```diff
@@ -0,0 +1,260 @@
+if is_torch_available():
+    import torch
+
+if find_spec("jieba") is not None:
```
solved

younesbelkada commented:
Hi @gongbaitao 
Great work on refactoring the code and making the CI tests pass! 🎉 
Let me know once this is ready for review!

gongbaitao committed:
Merge remote-tracking branch 'upstream/main' into cpmantmodel

younesbelkada commented:
Hello @gongbaitao 
Don't worry I think this is fine. If you give me the green light, I can review the PR now

younesbelkada commented on src/transformers/models/cpmant/configuration_cpmant.py:
```diff
@@ -0,0 +1,129 @@
+logger = logging.get_logger(__name__)
+
+CPMANT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "cpm-ant-10b": "https://huggingface.co/cpm-ant-10b/resolve/main/config.json",
```
This checkpoint seems to be not public, can you please double check?, maybe you mean
`https://huggingface.co/openbmb/cpm-ant-10b/blob/main/config.json`

gongbaitao replied:
Thanks for the reminder! The ckpt has been public in https://huggingface.co/openbmb/cpm-ant-10b

younesbelkada commented on src/transformers/models/cpmant/configuration_cpmant.py:
```diff
@@ -0,0 +1,129 @@
+    This is the configuration class to store the configuration of a [`~CPMAntModel`]. It is used to instantiate an
+    CPMAnt model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the CPMAnt
+    [cpm-ant-10b](https://huggingface.co/cpm-ant-10b) architecture.
```
```suggestion
    [cpm-ant-10b](https://huggingface.co/openbmb/cpm-ant-10b) architecture.
```

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,970 @@
+]
+
+
+# Copied from transformers.models.bert.modeling_bert.load_tf_weights_in_bert with BERT -> load_tf_weights_in_cpmant
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.load_tf_weights_in_bert with bert -> cpmant
```

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,970 @@
+                If set to `True`, `past_key_values` key value states are returned and can be used to speed up decoding
+                (see `past_key_values`).
+        """
+        current_key_value = None
+        hidden_states = self.self_att(
+            hidden_states,
+            attention_mask=attention_mask,
+            position_bias=position_bias,
+            output_attentions=output_attentions,
+            past_key_values=past_key_values,
+            use_cache=use_cache,
+        )
+
+        hidden_states, attn_weights, current_key_value = hidden_states
```
```suggestion
        hidden_states = self.self_att(
            hidden_states,
            attention_mask=attention_mask,
            position_bias=position_bias,
            output_attentions=output_attentions,
            past_key_values=past_key_values,
            use_cache=use_cache,
        )

        hidden_states, attn_weights, current_key_value = hidden_states
```

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,970 @@
+    def __init__(
+        self,
+        config: CPMAntConfig,
+        bidirectional: bool = True,
```
Is this always set to `True`? In this case no need to hardcode it no?

gongbaitao replied:
Yes it's always `True` in cpmant. I have dropped this hardcode.

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,970 @@
+
+        # add prompts ahead
+        if input_ids.dtype != torch.int32:
+            input_ids = input_ids.int()
```
```suggestion
            input_ids = input_ids.to(torch.int32)
```
nit

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
````diff
@@ -0,0 +1,970 @@
+        Example:
+
+        Text Generation with CPMAntForCausalLM.
+        ```python
+        >>> from transformers import CPMAntTokenizer, CPMAntForCausalLM
+
+        >>> texts = "今天天气不错，"
+        >>> model = CPMAntForCausalLM.from_pretrained("openbmb/cpm-ant-10b")
+        >>> tokenizer = CPMAntTokenizer.from_pretrained("openbmb/cpm-ant-10b")
+        >>> input_ids = tokenizer(texts, return_tensors="pt")
+        >>> outputs = model.generate(**input_ids)
+        >>> output_texts = tokenizer.batch_decode(outputs)
+        >>> print(output_texts)
+        ['今天天气不错，阳光明媚，我和妈妈一起去超市买东西。\n在超市里，我看到了一个很好玩的玩具，它的名字叫“机器人”。它有一个圆圆的脑袋，两只圆圆的眼睛，还有一个圆圆的']
+        ```
````
We might need to skip this doc-test, is there a way to keep it on the docstring and skip it on the doc-test runner side @ydshieh ? 🙏

younesbelkada commented on src/transformers/models/cpmant/tokenization_cpmant.py:
```diff
@@ -0,0 +1,271 @@
+
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "cpm-ant-10b": "https://huggingface.co/cpm-ant-10b/resolve/main/vocab.txt",
```
```suggestion
        "cpm-ant-10b": "https://huggingface.co/openbmb/cpm-ant-10b/resolve/main/vocab.txt",
```

younesbelkada commented on src/transformers/models/cpmant/tokenization_cpmant.py:
```diff
@@ -0,0 +1,271 @@
+        unk_token (`str`, *optional*, defaults to `<|endoftext|>`):
+            The unknown token.
+        line_token (`str`):
+            the line token.
+        space_token (`str`):
+            the space token.
```
```suggestion
            The line token.
        space_token (`str`, *optional*, defaults to `</_>`):
            The space token.
```

younesbelkada commented on src/transformers/models/cpmant/tokenization_cpmant.py:
```diff
@@ -0,0 +1,271 @@
+            The token used for padding.
+        unk_token (`str`, *optional*, defaults to `<|endoftext|>`):
+            The unknown token.
+        line_token (`str`):
```
```suggestion
        line_token (`str`, *optional*, defaults to `</n>`):
```

younesbelkada commented on src/transformers/models/cpmant/tokenization_cpmant_fast.py:
```diff
@@ -0,0 +1,293 @@
+
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "openbmb/cpm-ant-10b": "https://huggingface.co/cpm-ant-10b/resolve/main/vocab.txt",
+    },
+    "tokenizer_file": {
+        "openbmb/cpm-ant-10b": "https://huggingface.co/cpm-ant-10b/resolve/main/tokenizer.json",
```
```suggestion
        "openbmb/cpm-ant-10b": "https://huggingface.co/openbmb/cpm-ant-10b/resolve/main/vocab.txt",
    },
    "tokenizer_file": {
        "openbmb/cpm-ant-10b": "https://huggingface.co/openbmb/cpm-ant-10b/resolve/main/tokenizer.json",
```

younesbelkada commented on src/transformers/models/cpmant/tokenization_cpmant_fast.py:
```diff
@@ -0,0 +1,293 @@
+            The token used for padding.
+        unk_token (`str`, *optional*, defaults to `<|endoftext|>`):
+            The unknown token.
+        line_token (`str`):
+            the line token.
+        space_token (`str`):
+            the space token.
```
```suggestion
        line_token (`str`, *optional*, defaults to `</n>`):
            the line token.
        space_token (`str`, *optional*, defaults to `</_>`):
            the space token.
```

younesbelkada commented on tests/models/cpmant/test_modeling_cpmant.py:
```diff
@@ -0,0 +1,251 @@
+        self.assertEqual(expected_output, output_texts)
+
+    @slow
+    def test_batch_generation(self):
```
```suggestion
    @unittest.skip("skip this test as the model is very large for our daily runner")
    def test_batch_generation(self):
```
please add this on all `slow` tests here 🙏

younesbelkada commented on tests/models/cpmant/test_tokenization_cpmant.py:
```diff
@@ -0,0 +1,42 @@
+
+@custom_tokenizers
+class CPMAntTokenizationTest(TokenizerTesterMixin, unittest.TestCase):
+    def test_pre_tokenization(self):
```
can you add the same test for `CPMAntTokenizerFast` ?

sgugger replied:
Still needs to be addressed.

younesbelkada reviewed (commented):
This looks great to me! I left few comments, we should be close merging this after addressing the last comments 
I left only minor comments, let's add tokenizer tests for the slow tokenizer
I am ok for the naming (as in transformers we prefer camel-cased naming) in principle that we adopted the same approach as on this PR for `GPTSanJapanese`, so no problem for me to keep it `CPMAnt`.
My main comment being: please make sure to use the correct link for the checkpoints you mention on the codebase

Thanks a lot for the great work on this!

gongbaitao committed:
adjust code to some specifications

gongbaitao committed:
make style and fix-copies

gongbaitao committed:
Merge remote-tracking branch 'upstream/main' into cpmantmodel

gongbaitao committed:
add fast tokenization test

gongbaitao committed:
Merge remote-tracking branch 'upstream/main' into cpmantmodel

gongbaitao committed:
dummy commit for unit test

gongbaitao committed:
dummy commit for unit test

gongbaitao committed:
dummy commit for unit test

gongbaitao commented:
Hi @younesbelkada, thanks for your meaningful comments!

1. The link for checkpoint, and some other comments problems, have been corrected.

2. As for the name, because till now CPMAnt has just one Chinese case, I guess it's no need to call it like CPMAntChinese. 

3. Besides, I have made some local tests on tokenizers, for example, the `CPMAntTokenizationTest.test_pre_tokenization()`. But some methods in `TokenizerTesterMixin` use different logic to load vocab as a `dict`, while CPMAnt has it's own `load_vocab`. Refactor is not that convenient and necessary i think, so I just set it as `custom`.  I can make a change if there's any better solutions.

It's ok for the new review : )

younesbelkada commented on src/transformers/models/cpmant/configuration_cpmant.py:
```diff
@@ -0,0 +1,129 @@
+            just in case (e.g., 512 or 1024 or 2048).
+        eps (`float`, *optional*, defaults to 1e-6):
+            The epsilon used by the layer normalization layers.
+        half (`bool`, *optional*, defaults to False):
+            The dtype of tensor.
```
```suggestion
```
This is not used

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,964 @@
+        return hidden_states, current_key_values, all_hidden_states, all_self_attns
+
+
+# Copied from transformers.models.bert.modeling_bert.BertIntermediate with bert -> cpmant
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertIntermediate with Bert->CPMAnt
```

younesbelkada replied:
you need to remove the spaces !

younesbelkada replied:
Also make sure to use `Bert` & `CPMAnt` and not `bert` & `cpmant`

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,964 @@
+]
+
+
+# Copied from transformers.models.bert.modeling_bert.load_tf_weights_in_bert with bert -> cpmant
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.load_tf_weights_in_bert with bert->cpmant
```

younesbelkada replied:
same here, you need these spaces to be removed

younesbelkada commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,964 @@
+        return relative_buckets
+
+
+# Copied from transformers.models.bert.modeling_bert.BertOutput with bert -> cpmant
```
```suggestion
# Copied from transformers.models.bert.modeling_bert.BertOutput with Bert->CPMAnt
```

younesbelkada commented on tests/models/cpmant/test_modeling_cpmant.py:
```diff
@@ -0,0 +1,257 @@
+    )
+
+
+class CPMAntConfigTester(ConfigTester):
+    def create_and_test_config_common_properties(self):
+        config = self.config_class(**self.inputs_dict)
+        self.parent.assertTrue(hasattr(config, "dim_model"))
+        self.parent.assertTrue(hasattr(config, "num_heads"))
+        self.parent.assertTrue(hasattr(config, "dim_head"))
```
Can you explain why you need to ovveride the `ConfigTester` here? Check: https://github.com/huggingface/transformers/blob/dcec3277cdbbeafb6107e78f87d701111196c78a/tests/models/opt/test_modeling_opt.py#L212

gongbaitao replied:
Because our original model config has some different names. Actually it's not necessary, and I have renamed the arguments.

younesbelkada reviewed (commented):
Thanks a lot for this great addition! 
I left few final last comments, apart from that LGTM!

gongbaitao committed:
normalize some comments and names

gongbaitao committed:
Merge remote-tracking branch 'upstream/main' into cpmantmodel

gongbaitao committed:
Bert->CPMAnt

gongbaitao commented:
Hi @younesbelkada . I have fixed  the problems mentioned in comments. I think it's ready for new review:)
Thanks for your detailed review and comments!

younesbelkada commented on src/transformers/utils/dummy_pt_objects.py:
```diff
@@ -1755,6 +1755,34 @@ def __init__(self, *args, **kwargs):
+
+
+class CPMAntPreTrainedModel(metaclass=DummyObject):
+    _backends = ["torch"]
+
+    def __init__(self, *args, **kwargs):
+        requires_backends(self, ["torch"])
```
```suggestion
    _backends = ["torch", "jieba"]

    def __init__(self, *args, **kwargs):
        requires_backends(self, ["torch", "jieba"])
```

younesbelkada replied:
As I am not sure here, please wait for the review of @sgugger  🙏

sgugger replied:
The model has no requirement on jieba.

younesbelkada commented on src/transformers/utils/dummy_tokenizers_objects.py:
```diff
@@ -93,6 +93,13 @@ def __init__(self, *args, **kwargs):
 
 
+class CPMAntTokenizerFast(metaclass=DummyObject):
+    _backends = ["tokenizers"]
+
+    def __init__(self, *args, **kwargs):
+        requires_backends(self, ["tokenizers"])
+
```
```suggestion
    _backends = ["tokenizers", "jieba"]

    def __init__(self, *args, **kwargs):
        requires_backends(self, ["tokenizers", "jieba"])

```

younesbelkada reviewed (approved):
Thanks a lot for this huge work 💪 This looks great to me!
I left two open questions, let's see what @sgugger will say! 
Thanks again for your dedication

sgugger commented on src/transformers/models/cpmant/configuration_cpmant.py:
```diff
@@ -0,0 +1,127 @@
+
+class CPMAntConfig(PretrainedConfig):
+    r"""
+    This is the configuration class to store the configuration of a [`~CPMAntModel`]. It is used to instantiate an
```
```suggestion
    This is the configuration class to store the configuration of a [`CPMAntModel`]. It is used to instantiate an
```

sgugger commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,965 @@
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "cpm-ant-10b"
```
This is not a valid checkpoint name.

sgugger commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,965 @@
+
+        self.eps = config.eps
+        self.dim_norm = config.hidden_size
+        self.weight = torch.nn.parameter.Parameter(torch.empty(config.hidden_size))
```
```suggestion
        self.weight = nn.Parameter(torch.empty(config.hidden_size))
```

sgugger commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,965 @@
+
+    @add_start_docstrings_to_model_forward(CPMANT_INPUTS_DOCSTRING)
+    @add_code_sample_docstrings(
+        processor_class=_TOKENIZER_FOR_DOC,
```
```suggestion
```

sgugger commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,965 @@
+
+    @add_start_docstrings_to_model_forward(CPMANT_INPUTS_DOCSTRING)
+    @add_code_sample_docstrings(
+        processor_class=_TOKENIZER_FOR_DOC,
```
```suggestion
```

sgugger commented on src/transformers/models/cpmant/tokenization_cpmant.py:
```diff
@@ -0,0 +1,271 @@
+            padding_side="left",
+            **kwargs,
+        )
+        self.bod_token = bod_token
```
There should be a `requires_backends["jieba"]` here, so that there is an error early if the dep is not installed.

gongbaitao replied:
Added. `requires_backend(self, ["jieba])`

sgugger commented on src/transformers/models/cpmant/tokenization_cpmant_fast.py:
```diff
@@ -0,0 +1,290 @@
+        return sub_tokens
+
+
+class CPMAntTokenizerFast(PreTrainedTokenizer):
```
This is not a fast tokenizer at all as it doesn't use the tokenizers library. This whole file should be removed.

sgugger reviewed (commented):
Thanks for adding this model. There are a few problems with the implementation before we can merge.

1. On the naming, we only use fully camel-cased names now, so the classes should be CpmAntConfig, CpmAntModel etc.
2. The fast tokenizer is not fast at all. It doesn't use the tokenizers library, not does it inherit from the right base class. It's fine to not have a fast tokenizer, the class and the corresponding files should be removed.

younesbelkada commented on tests/models/cpmant/test_modeling_cpmant.py:
```diff
@@ -0,0 +1,249 @@
+@require_torch
+class CPMAntModelIntegrationTest(unittest.TestCase):
+    @slow
+    @unittest.skip("skip this test as the model is very large for our daily runner")
```
Here we meant to completely skip the test as it will not fit the GPU RAM of the CI runners, this checkpoint is 20GB in half-precision

gongbaitao replied:
Solved : )

sgugger replied:
Corresponding changes here do not seem to have been pushed.

gongbaitao committed:
camel names and drop redundant codes

gongbaitao committed:
Merge remote-tracking branch 'upstream/main' into cpmantmodel

gongbaitao committed:
make style and fix-coies

gongbaitao commented on docs/source/en/model_doc/cpmant.mdx:
```diff
@@ -0,0 +1,49 @@
+⚙️ Training & Inference
+- A tutorial on [CPM-Live](https://github.com/OpenBMB/CPM-Live/tree/cpm-ant/cpm-live).
+
+## CPMAntConfig
```
All corrected : )

gongbaitao commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,965 @@
+        return hidden_states, attn_weights, current_key_value
+
+
+class DenseGatedACT(nn.Module):
```
Added

gongbaitao committed:
add CpmTokenizerFast _import_structure

gongbaitao committed:
drop cpmanttokenizerfast in model_doc

gongbaitao commented:
@younesbelkada @sgugger Thanks for the valued comments!
According to the new comments, I have dropped some redundant codes, and rename the model class in a camel-cased way : )

sgugger commented on README.md:
```diff
@@ -309,6 +309,7 @@ Current number of checkpoints: ![](https://img.shields.io/endpoint?url=https://h
 1. **[ConvBERT](https://huggingface.co/docs/transformers/model_doc/convbert)** (from YituTech) released with the paper [ConvBERT: Improving BERT with Span-based Dynamic Convolution](https://arxiv.org/abs/2008.02496) by Zihang Jiang, Weihao Yu, Daquan Zhou, Yunpeng Chen, Jiashi Feng, Shuicheng Yan.
 1. **[ConvNeXT](https://huggingface.co/docs/transformers/model_doc/convnext)** (from Facebook AI) released with the paper [A ConvNet for the 2020s](https://arxiv.org/abs/2201.03545) by Zhuang Liu, Hanzi Mao, Chao-Yuan Wu, Christoph Feichtenhofer, Trevor Darrell, Saining Xie.
 1. **[CPM](https://huggingface.co/docs/transformers/model_doc/cpm)** (from Tsinghua University) released with the paper [CPM: A Large-scale Generative Chinese Pre-trained Language Model](https://arxiv.org/abs/2012.00413) by Zhengyan Zhang, Xu Han, Hao Zhou, Pei Ke, Yuxian Gu, Deming Ye, Yujia Qin, Yusheng Su, Haozhe Ji, Jian Guan, Fanchao Qi, Xiaozhi Wang, Yanan Zheng, Guoyang Zeng, Huanqi Cao, S ...
+1. **[CpmAnt](https://huggingface.co/docs/transformers/main/model_doc/cpmant)** (from OpenBMB) released by the [OpenBMB](https://www.openbmb.org/).
```
```suggestion
1. **[CPMAnt](https://huggingface.co/docs/transformers/main/model_doc/cpmant)** (from OpenBMB) released by the [OpenBMB](https://www.openbmb.org/).
```
This should use the model name casing. It's only the model/config/tokenizer classes that should be CpmAntXxx

gongbaitao replied:
Solved.

sgugger commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -402,6 +404,7 @@
         ("convbert", "ConvBERT"),
         ("convnext", "ConvNeXT"),
         ("cpm", "CPM"),
+        ("cpmant", "CpmAnt"),
```
```suggestion
        ("cpmant", "CPM-Ant"),
```

gongbaitao replied:
Solved.

sgugger commented on src/transformers/models/cpmant/configuration_cpmant.py:
```diff
@@ -0,0 +1,127 @@
+logger = logging.get_logger(__name__)
+
+CPMANT_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "cpm-ant-10b": "https://huggingface.co/openbmb/cpm-ant-10b/blob/main/config.json"
```
```suggestion
    "openbmb/cpm-ant-10b": "https://huggingface.co/openbmb/cpm-ant-10b/blob/main/config.json"
```

gongbaitao replied:
Solved.

sgugger commented on src/transformers/models/cpmant/configuration_cpmant.py:
```diff
@@ -0,0 +1,127 @@
+    This is the configuration class to store the configuration of a [`CpmAntModel`]. It is used to instantiate an
+    CPMAnt model according to the specified arguments, defining the model architecture. Instantiating a configuration
+    with the defaults will yield a similar configuration to that of the CPMAnt
+    [cpm-ant-10b](https://huggingface.co/openbmb/cpm-ant-10b) architecture.
```
```suggestion
    [openbmb/cpm-ant-10b](https://huggingface.co/openbmb/cpm-ant-10b) architecture.
```

gongbaitao replied:
Solved.

sgugger commented on src/transformers/models/cpmant/tokenization_cpmant.py:
```diff
@@ -0,0 +1,272 @@
+
+PRETRAINED_VOCAB_FILES_MAP = {
+    "vocab_file": {
+        "cpm-ant-10b": "https://huggingface.co/openbmb/cpm-ant-10b/blob/main/vocab.txt",
```
```suggestion
        "openbmb/cpm-ant-10b": "https://huggingface.co/openbmb/cpm-ant-10b/blob/main/vocab.txt",
```

gongbaitao replied:
Solved.

sgugger commented on src/transformers/models/cpmant/tokenization_cpmant.py:
```diff
@@ -0,0 +1,272 @@
+}
+
+PRETRAINED_POSITIONAL_EMBEDDINGS_SIZES = {
+    "cpm-ant-10b": 1024,
```
```suggestion
    "openbmb/cpm-ant-10b": 1024,
```

gongbaitao replied:
Solved.

sgugger commented on src/transformers/models/cpmant/tokenization_cpmant.py:
```diff
@@ -0,0 +1,272 @@
+    Args:
+        vocab_file (`str`):
+            Path to the vocabulary file.
+        bod_token (`str`, *optional*, defaults to `<d>`):
+            The beginning of document token.
+        eod_token (`str`, *optional*, defaults to `</d>`):
+            The end of document token.
+        bos_token (`str`, *optional*, defaults to `<s>`):
+            The beginning of sequence token.
+        eos_token (`str`, *optional*, defaults to `</s>`):
+            The end of sequence token.
+        pad_token (`str`, *optional*, defaults to `<pad>`):
+            The token used for padding.
+        unk_token (`str`, *optional*, defaults to `<unk>`):
+            The unknown token.
+        line_token (`str`, *optional*, defaults to `</n>`):
+            The line token.
+        space_token (`str`, *optional*, defaults to `</_>`):
```
Default values are all missing double quotes here.

gongbaitao replied:
Solved.

sgugger commented on tests/models/cpmant/test_tokenization_cpmant.py:
```diff
@@ -0,0 +1,42 @@
+    from transformers.models.cpmant import CpmAntTokenizer
+
+
+@unittest.skip("CPMAntTokenizer process vocab in list format, so we skip the common test.")
```
The test below won't be executed because of this global skip here.

gongbaitao replied:
I skip the test because it's `load_vocab` logic is different from the `TokenizerTesterMixin`. While refactor is not that convenient and necessary i think, so I just skip it. Is it OK? Need I make some changes?

sgugger replied:
No this is not ok, since the test you wrote below will never be executed. If all tests fail in the TokenizerTesterMixin, that means your tokenizer does not have an API consistent with the other tokenizers of Transformers, and thus we can't accept it. You should fix your tokenizer so that it passes most of the tests of the common tester.

gongbaitao replied:
OK I will fix it, thanks!

gongbaitao replied:
> The test below won't be executed because of this global skip here.

Solved.

sgugger reviewed (commented):
Thanks for iterating! There are still a couple of issues here and there. Also the test added for the tokenizer will need to be decorated with a `requires_jieba` (that you will need to define in testing_utils, similar to the other requires_xxx functions). Lastly, you need to also add an import error for jieba in `import_utils.py` so that `requires_backend(["jieba"])` works without error.

gongbaitao committed:
fix some problems

gongbaitao committed:
fix CPMAnt tokenization for common test

gongbaitao committed:
solve conflict with upstream/main

gongbaitao committed:
make style and fixup

gongbaitao committed:
fix copies and fixup

gongbaitao committed:
fix bugs in tokenization test

gongbaitao committed:
dummy commit for connection failure in unittest

gongbaitao commented:
Hi @sgugger @younesbelkada , sorry for the delay!
In the last few weeks, I have fixed the problems mentioned above and refactored the CPMAnt tokenizer. Please kindly have a look again, thanks for your help!

sgugger commented on tests/models/cpmant/test_modeling_cpmant.py:
```diff
@@ -0,0 +1,249 @@
+@require_torch
+class CPMAntForCausalLMlIntegrationTest(unittest.TestCase):
+    @slow
+    @unittest.skip("skip this test as the model is very large for our daily runner")
```
Same here.

sgugger commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,886 @@
+    We use Root Mean Square (RMS) Layer Normalization, please see https://arxiv.org/abs/1910.07467 for details."
+    """
+
+    def __init__(
+        self,
+        config: CpmAntConfig,
+    ):
```
```suggestion
    def __init__(self, config: CpmAntConfig):
```

gongbaitao replied:
It seems to be result of `make style`. After I fits the code in one line, I cannot pass the 
CI check in code quality.

sgugger replied:
That's because you are not copy-pasting this suggestion as is (there is a button to accept it directly in GitHub FYI) but are leaving a trailing comma.

gongbaitao replied:
Sorry I misunderstand the problem as fits in one line. Now the trailing comma has beed fixed:)

sgugger commented on src/transformers/models/cpmant/modeling_cpmant.py:
```diff
@@ -0,0 +1,886 @@
+
+
+class CpmAntSegmentPositionEmbedding(nn.Module):
+    def __init__(
+        self,
+        config: CpmAntConfig,
+    ):
```
```suggestion
    def __init__(self, config: CpmAntConfig):
```

sgugger reviewed (commented):
Thanks for iterating! We're almost good to go, just a couple of comments left to address in the tests.

gongbaitao commented:
Thanks for your quick review! @sgugger 
It seems this problem https://github.com/huggingface/transformers/pull/20906#discussion_r1161687724 is because the changed file didn't show all commits. Maybe check this page https://github.com/huggingface/transformers/pull/20906/files will be helpful:)
As the https://github.com/huggingface/transformers/pull/20906#discussion_r1161688147, it cannot pass the code quality check, so shall I keep it unchanged?

gongbaitao committed:
solve conflict

gongbaitao committed:
fix copies

sgugger reviewed (commented):
Replied on the comment for the styling issue. The exact same line is present multiple times in the modeling file so I think you did not take the suggestion as it is written.

As for the tests, i'm sorry I was unclear: I meant that we should use the `@tooslow` decorator instead of `skip`.

gongbaitao committed:
drop trailing comma

gongbaitao committed:
fix decorator in tests

gongbaitao committed:
Merge remote-tracking branch 'upstream/main' into cpmantmodel

gongbaitao committed:
dummy commit for connection failure in unittest

sgugger reviewed (approved):
Perfect, thanks for bearing with me :-)
Congrats on getting this new model merged into Transformers and thanks again for all your work!
